libvirt/tunnel/tunnel.sh

```diff
@@ -1,98 +1,163 @@
-# echo "-R ${CLUSTER_PORTS['2,API']}:127.0.0.1:26443  -R ${CLUSTER_PORTS['2,HTTP']}:127.0.0.1:20080 -R ${CLUSTER_PORTS['2,HTTPS']}:127.0.0.1:20443"
-# echo "-R ${CLUSTER_PORTS['3,API']}:127.0.0.1:36443  -R ${CLUSTER_PORTS['3,HTTP']}:127.0.0.1:30080 -R ${CLUSTER_PORTS['3,HTTPS']}:127.0.0.1:30443"
-# echo "-R ${CLUSTER_PORTS['4,API']}:127.0.0.1:46443  -R ${CLUSTER_PORTS['4,HTTP']}:127.0.0.1:40080 -R ${CLUSTER_PORTS['4,HTTPS']}:127.0.0.1:40443"
-
-function OC() {
-	./oc --server https://api.ci.openshift.org --token "${token}" --namespace "${bastion_name}" "${@}"
+
+function OC() {	./oc --server https://api.ci.openshift.org --token "${token}" --namespace "${environment}" "${@}"
+}
+
+function timestamp() {
+	# With UTC format.  2020-11-04 06:19:24
+	date -u  +"%Y-%m-%d %H:%M:%S"
 }
 
 function port-forward() {
 	while true; do
-		echo "[INFO] Setting up port-forwarding..."
+		echo "$(timestamp) [INFO] Setting up port-forwarding to connect to the bastion..."
 		pod="$( OC get pods --selector component=sshd -o jsonpath={.items[0].metadata.name} )"
-		if ! OC port-forward "${pod}" 2222; then
-			echo "[WARNING] Port-forwarding failed, retrying..."
+		if ! OC port-forward "${pod}" "${1:?Port was not specified}"; then
+			echo "$(timestamp) [WARNING] Port-forwarding failed, retrying..."
 			sleep 0.1
 		fi
 	done
 }
 
+# This opens an ssh tunnel. It uses port 2222 for the ssh traffic.
+# It basically says send traffic from bastion service port $1 to
+# local VM port $1 using port 2222 to establish the ssh connection.
 function ssh-tunnel() {
 	while true; do
-		echo "[INFO] Setting up SSH tunnelling..."
-		if ! ssh -N -T root@127.0.0.1 -p 2222 \
-                        -R ${BASTION_PORTS["LIBVIRT"]}:127.0.0.1:16509 \
-                        -R ${CLUSTER_PORTS["0,API"]}:127.0.0.1:6443   -R ${CLUSTER_PORTS["0,HTTP"]}:127.0.0.1:80    -R ${CLUSTER_PORTS["0,HTTPS"]}:127.0.0.1:443 \
-                        -R ${CLUSTER_PORTS["1,API"]}:127.0.0.1:16443  -R ${CLUSTER_PORTS["1,HTTP"]}:127.0.0.1:10080 -R ${CLUSTER_PORTS["1,HTTPS"]}:127.0.0.1:10443 \
-                        -R ${CLUSTER_PORTS["2,API"]}:127.0.0.1:26443  -R ${CLUSTER_PORTS["2,HTTP"]}:127.0.0.1:20080 -R ${CLUSTER_PORTS["2,HTTPS"]}:127.0.0.1:20443 \
-                        -R ${CLUSTER_PORTS["3,API"]}:127.0.0.1:36443  -R ${CLUSTER_PORTS["3,HTTP"]}:127.0.0.1:30080 -R ${CLUSTER_PORTS["3,HTTPS"]}:127.0.0.1:30443 \
-                        -R ${CLUSTER_PORTS["4,API"]}:127.0.0.1:46443  -R ${CLUSTER_PORTS["4,HTTP"]}:127.0.0.1:40080 -R ${CLUSTER_PORTS["4,HTTPS"]}:127.0.0.1:40443 \
-		; then
-			echo "[WARNING] SSH tunnelling failed, retrying..."
+		echo "$(timestamp) [INFO] Setting up a reverse SSH tunnel to connect bastion port "${1:?Bastion service port was not specified}" to VM port "${1:?VM port was not specified}"..."
+		if ! ssh -N -T root@127.0.0.1 -p 2222 -R "$1:127.0.0.1:$1"; then
```

Comment:
This won't work because there's no guarantee that the BASTION_PORT == LOCAL_PORT
This is why we set up the lookup tables. CLUSTER_PORTS & BASTION_PORTS are both lookup tables that represent what the port # should be on the bastion pod.

Example:  ${CLUSTER_PORTS["0,HTTP"]}:127.0.0.1:80 ==> BASTION_PORT(8080):127.0.0.1:LOCAL_PORT(80) LPAR1, but BASTION_PORT(8081):127.0.0.1:LOCAL_PORT(80) LPAR2